torch/testing/asserts.py

```diff
@@ -0,0 +1,382 @@
+
+
+def _check_supported(a: torch.Tensor, b: torch.Tensor) -> None:
+    if any(t.dtype in (torch.complex32, torch.complex64, torch.complex128) for t in (a, b)):
+        raise UsageError("Comparison for complex tensors is not supported yet.")
+    if any(t.is_quantized for t in (a, b)):
+        raise UsageError("Comparison for quantized tensors is not supported yet.")
+    if any(t.is_sparse for t in (a, b)):
+        raise UsageError("Comparison for sparse tensors is not supported yet.")
+
+
+def _assert_attributes_equal(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    /,
+    *,
+    check_device: bool = True,
+    check_dtype: bool = True,
+    check_stride: bool = True,
+) -> None:
+    """Asserts that attributes of two tensors match.
+
+    Always checks the :attr:`~torch.Tensor.shape`. Other checks are optional and can be disabled.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+        check_device (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` live in the same
+            :attr:`~torch.Tensor.device` memory.
+        check_dtype (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :attr:`~torch.Tensor.dtype`.
+        check_stride (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :meth:`~torch.Tensor.stride`.
+
+    Raises:
+        AssertionError: If :attr:`a` and :attr:`b` do not have the same :attr:`~torch.Tensor.shape`.
+        AssertionError: If :attr:`check_device`, but :attr:`a` and :attr:`b` do not live in the same
+            :attr:`~torch.Tensor.device` memory.
+        AssertionError: If :attr:`check_dtype`, but :attr:`a` and :attr:`b` do not have the same
+            :attr:`~torch.Tensor.dtype`.
+        AssertionError: If :attr:`check_stride`, but :attr:`a` and :attr:`b` do not have the same
+            :meth:`~torch.Tensor.stride`.
+    """
+    msg_fmtstr = "The values for attribute '{}' do not match: {} != {}."
+
+    if a.shape != b.shape:
+        raise AssertionError(msg_fmtstr.format("shape", a.shape, b.shape))
+
+    if check_device and a.device != b.device:
+        raise AssertionError(msg_fmtstr.format("device", a.device, b.device))
+
+    if check_dtype and a.dtype != b.dtype:
+        raise AssertionError(msg_fmtstr.format("dtype", a.dtype, b.dtype))
+
+    if check_stride and a.stride() != b.stride():
+        raise AssertionError(msg_fmtstr.format("stride()", a.stride(), b.stride()))
+
+
+def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Equalizes some attributes of two tensors for value comparison.
+
+    If :attr:`a` and :attr:`b`
+    - do not live in the same memory :attr:`~torch.Tensor.device`, they are moved CPU memory, and
+    - do not have the same :attr:`~torch.Tensor.dtype`, they are copied to the :class:`~torch.dtype` returned by
+        :func:`torch.promote_types`.
```

Comment:
Should we also invoke `.contiguous()` if the strides don't match?

Reply:
I don't think we need to, and we change that as an implementation detail later if we like